crates/matrix-sdk/src/notification_settings/mod.rs

```diff
@@ -12,3 +26,722 @@ pub enum RoomNotificationMode {
+pub struct NotificationSettings {
+    /// The underlying HTTP client.
+    client: Client,
+    /// Owner's account push rules. They will be updated on sync.
+    rules: Arc<RwLock<Rules>>,
+    /// Event handler for push rules event
+    push_rules_event_handler: EventHandlerHandle,
+}
+
+impl Drop for NotificationSettings {
+    fn drop(&mut self) {
+        self.client.remove_event_handler(self.push_rules_event_handler.clone());
+    }
+}
+
+impl NotificationSettings {
+    /// Build a new `NotificationSettings``
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - A `Client` used to perform API calls
+    /// * `ruleset` - A `Ruleset` containing account's owner push rules
+    pub fn new(client: Client, ruleset: Ruleset) -> Self {
+        let rules = Arc::new(RwLock::new(Rules::new(ruleset)));
+
+        // Listen for PushRulesEvent
+        let rules_clone = rules.clone();
+        let push_rules_event_handler = client.add_event_handler(move |ev: PushRulesEvent| {
+            let rules = rules_clone.clone();
+            async move {
+                *rules.write().await = Rules::new(ev.content.global);
+            }
+        });
+
+        Self { client, rules, push_rules_event_handler }
+    }
+
+    /// Gets all user defined rules matching a given `room_id`.
+    async fn get_custom_rules_for_room(&self, room_id: &RoomId) -> Vec<(RuleKind, String)> {
+        self.rules.read().await.get_custom_rules_for_room(room_id)
+    }
+
+    /// Gets the user defined notification mode for a room.
+    pub async fn get_user_defined_room_notification_mode(
+        &self,
+        room_id: &RoomId,
+    ) -> Option<RoomNotificationMode> {
+        self.rules.read().await.get_user_defined_room_notification_mode(room_id)
+    }
+
+    /// Gets the default notification mode for a room.
+    ///
+    /// # Arguments
+    ///
+    /// * `is_encrypted` - `true` if the room is encrypted
+    /// * `members_count` - the room members count
+    pub async fn get_default_room_notification_mode(
+        &self,
+        is_encrypted: bool,
+        members_count: u64,
+    ) -> RoomNotificationMode {
+        self.rules.read().await.get_default_room_notification_mode(is_encrypted, members_count)
+    }
+
+    /// Get whether the given ruleset contains some enabled keywords rules.
+    pub async fn contains_keyword_rules(&self) -> bool {
+        self.rules.read().await.contains_keyword_rules()
+    }
+
+    /// Get whether a push rule is enabled.
+    pub async fn is_push_rule_enabled(
+        &self,
+        kind: RuleKind,
+        rule_id: &str,
+    ) -> Result<bool, NotificationSettingsError> {
+        self.rules.read().await.is_enabled(kind, rule_id)
+    }
+
+    /// Set whether a push rule is enabled.
+    pub async fn set_push_rule_enabled(
+        &self,
+        kind: RuleKind,
+        rule_id: &str,
+        enabled: bool,
+    ) -> Result<(), NotificationSettingsError> {
+        let rules = self.rules.read().await.clone();
```

Comment:
nit: This takes the lock while the request is in flight in `execute_commands`, preventing other write operations; make sure to use a smaller scope for taking this lock, until it's not necessary anymore.

(Funnily, I think it works that you can get the `write()` lock below because you're reusing the same variable name, which drops the previous one just before the new binding. With a different name, I'd expect a deadlock to happen \o/)

Reply:
Here the lock isn't release after the clone operation?